SG14_test/slot_map_test.cpp

```diff
@@ -1,26 +1,36 @@
 #include "SG14_test.h"
 #include "slot_map.h"
-#include <assert.h>
-#include <inttypes.h>
 #include <algorithm>
+#include <assert.h>
 #include <deque>
 #include <forward_list>
-#include <list>
+#include <inttypes.h>
 #include <iterator>
+#include <list>
 #include <memory>
 #include <random>
 #include <type_traits>
 #include <utility>
+#include <vector>
 
 namespace TestKey {
 struct key_16_8_t {
     uint16_t index;
     uint8_t generation;
+    friend bool operator==(key_16_8_t lhs, key_16_8_t rhs);
 };
+bool operator==(key_16_8_t lhs, key_16_8_t rhs) {
+    return lhs.index == rhs.index && lhs.generation == rhs.generation;
+}
```

Comment:
Please move this friend function in-line (use the "hidden friend idiom"), and also please change its name to something other than `operator==`, just so that it's clear that it's not part of the expected API of a key type. I suggest `friend bool KeysAreEqual(key_16_8_t lhs, key_16_8_t rhs)`.

If you do that, and pull the `slot_map::key(const_iterator)` member function into its own PR (or equivalently, remove the `at_unchecked` part from this PR), I'm favorably disposed toward it.

Reply:
I'll remove `at_unchecked`, do you also want me to remove the `operator[]` changes? I'm quite opposed to that. The current subscript implementation is quite dangerous.

Reply:
Great blog post about ADL, is this a ploy to drive traffic to your website? lol!

When looking up the hidden friend idiom, I found this : http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1601r0.pdf

So the idiom just states to make the operator inline friend. If I understand your ADL post correctly, this limits the exposure of the operator to ADL only.

I am curious what is the problem case that this fixes, when the operator is exposed in the enclosing namespace (declared out-of-line)?